personal_finances/user/user_auth_exceptions.py

```diff
@@ -0,0 +1,118 @@
+from typing import Dict, Tuple, Type
+import logging
+
+LOGGER = logging.getLogger(__name__)
+
+
+class AuthorizationHeaderNotPresent(Exception):
+    pass
+
+
+class AuthorizationHeaderEmptyContent(Exception):
+    pass
+
+
+class InvalidAuthorizationHeader(Exception):
+    pass
+
+
+class InvalidLambdaEventInput(Exception):
+    pass
+
+
+class UserNotFound(Exception):
+    pass
+
+
+class PasswordNotMatch(Exception):
+    pass
+
+
+class InvalidUserPassword(Exception):
+    pass
+
+
+class InvalidUserName(Exception):
+    pass
+
+
+class InvalidDynamoResponse(Exception):
+    pass
+
+
+class InconsistentExceptionDictionary(Exception):
+    pass
+
+
+"""
+    This dicts contains the map between the exceptions generated in
+    the user login api and the http response. No exception should be
+    duplicated, otherwise the 'InconsistentExceptionDictionary' exception
+    will be raised in the '_get_exception_type_index' function
+"""
+_EXCEPTION_TO_HTTP_RESPONSE: Dict[Tuple, Dict] = {
+    (UserNotFound, PasswordNotMatch): {
+        "statusCode": 401,
+        "body": "User or password incorrect",
+    },
+    (
+        AuthorizationHeaderNotPresent,
+        AuthorizationHeaderEmptyContent,
+        InvalidAuthorizationHeader,
+    ): {
+        "statusCode": 400,
+        "body": "Invalid authentication input",
+    },
+    (
+        InvalidUserPassword,
+        InvalidUserName,
+    ): {
+        "statusCode": 400,
+        "body": "Username or password does not match the minimal security requirements",
+    },
+    (
+        InvalidLambdaEventInput,
+        InvalidDynamoResponse,
+        KeyError,
```

Comment:
That's quite a generic exception. KeyError's could be happening in many places for many different reasons. One case that makes the status code inconsistent is if the main code is parsing a malformed user input (currently or in future modifications to the main code).

I wouldn't map this intentionally to a 500. I'd rather catch a more specific exception like `InvalidEnvironmentVars`.

Reply:
Sure, I agree with not mapping such generic exceptions to 500 directly. Regarding your suggestion, is raising a KeyError exception as InvalidEnvironmentVars valid in trying to get env variables? In the end, both will lead to the same line in the traceback unless you plan to check the var's content consistency (for example, some pattern in the string value).

Reply:
You can either:
1. remove the `KeyError` from here and treat any `KeyError` as unknown.
2. replace the `KeyError` by `InvalidEnvVars` to map it explicitly.

I'm ok with either, both options will remove the assumption that `KeyError` is a 500 in all cases. The second option includes a bit more effort to parse env vars, and make it a bit more explicit.

As per your question, if you choose to add an exception called something like `InvalidEnvVars`, we should stick to its meaning, all cases you consider the environment variables as invalid you should use this exception. How low level you want to go it's up to you.